_posts/2022-08-08-understanding-meta-keyword-argument.md

```diff
@@ -0,0 +1,240 @@
+
+**Another note**, Dask Arrays may not always do this conversion. You can look at the API reference for your particular Array operation for details.
+
+```python
+>>> import numpy as np
+>>> import dask.array as da
+>>> my_arr = da.random.random(10)
+>>> my_arr.map_blocks(lambda x: len(x)).compute()
+10
+```
+
+## `meta` does not _force_ the structure or dtypes
+
+`meta` can be thought of as a suggestion to Dask. Dask uses this `meta` to generate the task graph until it can infer the actual metadata from the values. It **does not** force the output to have the structure or dtype of the specified `meta`.
+
+Consider the following example, and remember that we defined `ddf` with `x` and `y` column names in the previous sections.
+
+If we provide different column names (`a` and `b`) in the `meta` description, Dask uses these new names to create the task graph:
+
+```python
+>>> meta_df = pd.DataFrame(columns=["a", "b"], dtype=int)
+>>> result = ddf.apply(lambda x:x*x, axis=1, meta=meta_df)
+>>> result
+Dask DataFrame Structure:
+                   a      b
+npartitions=2
+0              int64  int64
+3                ...    ...
+5                ...    ...
+Dask Name: apply, 4 tasks
+```
+
+However, if we compute `result`, we will get the following error:
+
+```
+>>> result.compute()
+ValueError: The columns in the computed data do not match the columns in the provided metadata
+  Extra:   ['x', 'y']
+  Missing: ['a', 'b']
+```
+
+While computing, Dask evaluates the actual metadata with columns `x` and `y`. This does not match the `meta` that we provided, and hence, Dask raises a helpful error message. Notice how Dask does not change the output to have `a` and `b` here, rather uses `a` and `b` column names only for intermediate task graphs.
+
+## Using `_meta` directly
+
+In some rare case, you can also set the `_meta` attribute directly for a Dask DataFrame. For example, if the DataFrame was read with incorrect dtypes, like:
+
+```python
+>>> ddf = dd.DataFrame.from_dict(
+...     {
+...         "x": range(6),
+...         "y": range(10, 16),
+...     },
+...     dtype="object", # Note the “object” dtype
+...     npartitions=2,
+... )
+>>> ddf
+Dask DataFrame Structure:
+                    x       y
+npartitions=2
+0              object  object
+3                 ...     ...
+5                 ...     ...
+Dask Name: from_pandas, 2 tasks
+```
+
+The values are clearly integers but the dtype says `object`, so we can't perform integer operations like addition:
+
+```python
+>>> add = ddf + 2
+ValueError: Metadata inference failed in `add`.
+
+Original error is below:
+------------------------
+TypeError('can only concatenate str (not "int") to str')
+```
+
+Here, we can explicitly define `_meta`:
+
+```python
+>>> ddf._meta = pd.DataFrame(columns=["x", "y"], dtype="int64")
```

Comment:
_Not a blocker_. Again this feels strange because it is private, should we implement some public getters and setters for `_meta`?